sqlmesh/core/engine_adapter/base.py

```diff
@@ -107,43 +139,102 @@ def __init__(
         self.sql_gen_kwargs = sql_gen_kwargs or {}
         self._extra_config = kwargs
 
+    @property
+    def cursor(self) -> t.Any:
+        return self._connection_pool.get_cursor()
+
+    @property
+    def spark(self) -> t.Optional[PySparkSession]:
+        return None
+
     @classmethod
     def is_pandas_df(cls, value: t.Any) -> bool:
         return isinstance(value, pd.DataFrame)
 
     @classmethod
-    def is_pyspark_df(cls, value: t.Any) -> bool:
-        return hasattr(value, "sparkSession")
-
-    @classmethod
-    def is_df(self, value: t.Any) -> bool:
-        return self.is_pandas_df(value) or self.is_pyspark_df(value)
+    def _to_utc_timestamp(cls, col: t.Union[str, exp.Literal, exp.Column, exp.Null]) -> exp.Cast:
+        def ensure_utc_exp(
+            ts: t.Union[str, exp.Literal, exp.Column, exp.Null]
+        ) -> t.Union[exp.Literal, exp.Column, exp.Null]:
+            if not isinstance(ts, (str, exp.Literal)):
+                return ts
+            if isinstance(ts, exp.Literal):
+                if not ts.is_string:
+                    raise SQLMeshError("Timestamp literal must be a string")
+                ts = ts.this
+            return exp.Literal.string(
+                datetime.fromisoformat(t.cast(str, ts))
+                .astimezone(timezone.utc)
+                .strftime("%Y-%m-%d %H:%M:%S")
+            )
 
-    @classmethod
-    def try_get_df(cls, value: t.Any) -> t.Optional[DF]:
-        if cls.is_df(value):
-            return value
-        return None
+        return exp.cast(ensure_utc_exp(col), "TIMESTAMP")
 
-    @classmethod
-    def try_get_pyspark_df(cls, value: t.Any) -> t.Optional[PySparkDataFrame]:
-        if cls.is_pyspark_df(value):
-            return value
-        return None
+    def _get_source_queries(
+        self,
+        query_or_df: QueryOrDF,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]],
+        target_table: TableName,
+        *,
+        batch_size: t.Optional[int] = None,
+    ) -> t.List[SourceQuery]:
+        batch_size = self.DEFAULT_BATCH_SIZE if batch_size is None else batch_size
+        if isinstance(query_or_df, (exp.Subqueryable, exp.DerivedTable)):
+            return [SourceQuery(query_factory=lambda: query_or_df)]  # type: ignore
+        else:
+            return self._df_to_source_queries(
+                query_or_df, columns_to_types, batch_size, target_table=target_table
+            )
 
-    @classmethod
-    def try_get_pandas_df(cls, value: t.Any) -> t.Optional[pd.DataFrame]:
-        if cls.is_pandas_df(value):
-            return value
-        return None
+    def _df_to_source_queries(
+        self,
+        df: DF,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]],
+        batch_size: int,
+        target_table: TableName,
+    ) -> t.List[SourceQuery]:
+        assert isinstance(df, pd.DataFrame)
+        assert columns_to_types
```

Comment:
why are you using asserts here? if it's for typing why is it legal for these to be optional then?

Reply:
Only in the case of Spark PySparkDataframes can `columns_to_types` be None and that will be addressed in future PR.

Reply:
you should use overloading then i think, remember, asserts get removed at runtime